### DISCUSSION TITLE
Documenting py5 tips for those coming from Processing Java

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#389] [Ideas]
I have started this markdown document, I'd love to have more ideas and suggestions!

https://github.com/villares/processing-python/blob/main/migration-help/coming-from-Processing-Java.md

Some of the topics covered:

- The slightly different snake_case names
- How about the P-Classes?
   - PImage, PVector 
- About the libraries
- More tips for porting Processing Java code to py5
  - Getting started on quick code conversions!
  - A table with some equivalences for conversion
  - Looping with `for`  
  - Global variables
  - Strings & chars 
  - OO differences (and similarities)
  - Data structures

This started as issue https://github.com/py5coding/py5generator/issues/196, but I think it is better discussed here.

**Comment by hx2A** [comment]
What you wrote looked pretty complete so I tried adding it to the documentation page to see how it might fit:

http://dev.py5coding.org/content/coming_from_processing_java.html

First, are you OK with me adding your tips page to the py5coding website?

I made relatively minor changes to clean up the markdown a bit and spelling/grammar. Also, I changed the title to be shorter, made the py5coding links relative, and some other changes to make the markdown more consistent with the other pages. How do you think it looks?

I thought it best fit in the "Getting Started" section right after the Install page.

The raw markdown is here:

https://github.com/hx2A/py5book_fork/blob/dev/content/coming_from_processing_java.md

**Reply by villares** [reply]
Oh, I'm so glad you like it! Thank you for the copy editing! I'd love to see it on the official docs, the placement and changes seem great.

As you may have guessed, I have been working on versions of this for a long time, first as Processing Java to Python mode porting advice. I was wondering if we should have a page for "Python mode people" too. 

Let's see if we can fill in the last gaps, following Tom Larrow's comment I might want to add a pointer to type hints. Feel free to improve it anyway you see fit.

**Reply by hx2A** [reply]
> As you may have guessed, I have been working on versions of this for a long time, first as Processing Java to Python mode porting advice. I was wondering if we should have a page for "Python mode people" too.

Yes, you've been working hard on this, and I know the py5 community appreciates your hard work!

A Processing python mode page might be useful also. We could document the code conversion code built into py5.

> Let's see if we can fill in the last gaps, following Tom Larrow's comment I might want to add a pointer to type hints. Feel free to improve it anyway you see fit.

Sure, we can keep adding to it. Tom's comment is a good one - using type hints in Python is a good practice.

**Reply by villares** [reply]
I have mixed feelings about type hints, I'm kind of against using it too early with beginners (Cognitive Load Theory, etc.) and for exploratory coding, but I'll try to add something about it.

Another thing that came to my mind is we should add something about py5 modes and how this page will show examples in *imported mode* which is perhaps the most similar to Processing Java as written in the Processing IDE.

**Reply by hx2A** [reply]
> I have mixed feelings about type hints, I'm kind of against using it too early with beginners (Cognitive Load Theory, etc.) and for exploratory coding, but I'll try to add something about it.

I agree that type hints are not good for beginners but folks coming from Java are used to the compiler enforcing type safety. Before I started using Python I saw the lack of type safety as a negative feature of the language. People used to type safety might find it comforting to know that Python has optional type hints and that py5 uses them.

> Another thing that came to my mind is we should add something about py5 modes and how this page will show examples in imported mode which is perhaps the most similar to Processing Java as written in the Processing IDE.

Sure thing! Mentioning it useful, and also linking to the Imported Mode section of The Five py5 Modes would be a good idea.

http://dev.py5coding.org/content/py5_modes.html#imported-mode

Creating a link in Jupyter Book with the anchor a bit of a headache...I'll take care of that for you when you are ready for it.

**Reply by villares** [reply]
I created 2 PRs, but I have more things to fix... I'll try not to rush at it and enjoy my vacation a bit :)

**Reply by hx2A** [reply]
@villares yes, please enjoy your vacation! There's no rush for anything.

I just merged your 2 PRs and made a few small changes:

* used the `important` markdown directive at the top of the page
* converted the single character UTF ellipsis glyphs to three period characters (`…` to `...`). They look identical in some contexts but the former was tripping up the syntax highligher because it is not valid Java or Python code. Your text editor must be making the switch to that ellipsis glyph when you type 3 periods but that isn't a good idea for editing code.
* removed some unnecessary white space to make formatting consistent
* added a section towards the end for integer division (5 / 2 == 2 in Java but 2.5 in Python 3+)

You can see the rendered result on the dev website:

http://dev.py5coding.org/content/coming_from_processing_java.html

**Reply by hx2A** [reply]
Oh! I also re-worked the Install page.

http://dev.py5coding.org/content/install.html

I deprioritized everything having to do with Cairo and simplified things a bit. Also, made some updates that should have been made a while ago.

**Reply by villares** [reply]
I'm afraid that Apple now calls their OS [MacOS](https://support.apple.com/macos) maybe you could rename on the text but keep the URL for the pages so as not to break any links.

**Reply by hx2A** [reply]
> I'm afraid that Apple now calls their OS [MacOS](https://support.apple.com/macos) maybe you could rename on the text...

It seems I didn't get the memo - [they changed the name in 2016](https://en.wikipedia.org/wiki/MacOS#macOS).

But I am in good company - the Jupyter folks named the notebook event loop needed on MacOS for GUIs (including py5) "osx". They probably can't change it now but the could add an alias.

Anyhow, it is in the reference documentation all over the place and in some error messages. Fixing it is a simple search replace though.

> ...keep the URL for the pages so as not to break any links.

I can see from the log files that there are very few users "entering" the site from that page so renaming it doesn't have much of a downside. I am going to change that, I think it will be OK.

**Reply by hx2A** [reply]
And speaking of log files, I see that this month there were 5,905 visitors making 9,090 visits to py5coding.org. Last month was 4,097 visitors and 6,420 visits. Historically it has been around 2K visitors and 3-4K visits.

**Reply by hx2A** [reply]
The dev website is updated with the new (!) macOS terminology.

**Comment by villares** [comment]
Well, this discussion became a the source for https://py5coding.org/content/coming_from_processing_java.html
Thank you @hx2A!

If anyone wants to propose changes/additions the place to patch is https://github.com/py5coding/py5book/blob/main/content/coming_from_processing_java.md
